Mattermost auth integration

chenilim opened the pull request:
This allows Focalboard to authenticate via Mattermost, and provides a workspace per channel. To set up:
* Set `"authMode": "mattermost"` in config.json
* Set `"mattermostURL"` to your Mattermost test server's root url (e.g. "https://my.mattermost.com")
* Create an OAuth app in Mattermost, and populate `mattermostClientID` and `mattermostClientSecret`
* (If you have an existing install, clear your browser session, and delete the sessions table to avoid infinite loops - may have some to fix later)

To test:
* Get the channelID of a channel on your test server (from view channel info)
* Navigate to `http://localhost:8000/workspace/<channelID>`
* It should go through the OAuth loop and show an empty workspace
* Boards are scoped to each workspace

Breaking changes to the REST API:
* Add a `/workspaces/<workspaceID>/` prefix to the block APIs

We might need a few passes of reviews. To start, please review the overall login flow (in `mmauth.go`).

chenilim commented:
/check-cla

jwilander commented on server/mmauth/mmauth.go:
```diff
@@ -0,0 +1,249 @@
+	}
+
+	// Create session
+	// TODO: Remove any existing sessions with the same token? (Happens on re-login)
```
Token would change on re-login so that shouldn't be necessary

jwilander commented on server/mmauth/mmauth.go:
```diff
@@ -0,0 +1,249 @@
+	nativeUser, _ := a.store.GetUserById(user.Id)
+	if nativeUser == nil {
+		err = a.store.CreateUser(&model.User{
+			ID:          user.Id,
```
Directly mapping the focalboard user ID to the MM user ID is an interesting method. In the MM server we use the AuthData field for this and the user would get a unique focalboard ID too. This will work but it will tightly couple the user to the oauth service provider (in this case MM). If you ever wanted to change the auth service from MM to Google for example then this would become problematic since the focalboard user ID would need to change. Same problem if you wanted to support multiple login methods for one focalboard user.

jwilander replied:
Chatted with Chen in a DM about it. Leaving for now, may revisit in the future.

jwilander commented on server/mmauth/mmauth.go:
```diff
@@ -0,0 +1,249 @@
+	// TODO: Remove any existing sessions with the same token? (Happens on re-login)
+	session := model.Session{
+		ID:     uuid.New().String(),
+		Token:  token.AccessToken,
```
This is an interesting approach, using the the MM token as the authentication token for focalboard. My initial thought would be to create a focalboard specific session token and store the MM token in the session separately. Like the user ID above this will work but might get a little weird when using non-MM oauth service providers. I think this works more cleanly than the user ID above but you'd need to make sure that any requirements (eg length, characters) you have on the session token are loose enough to fit whatever pattern any oauth service provider might use for a token.

I can't think of any security reasons why this might be bad but @jupenur might have an opinion there

jupenur replied:
This is probably fine for MM auth but for generic OAuth I'd definitely create a new token.

jwilander commented on server/mmauth/mmauth.go:
```diff
@@ -0,0 +1,249 @@
+		Props: map[string]interface{}{
+			"authService":  "mattermost",
+			"expiry":       token.Expiry.Unix(),
+			"refreshToken": token.RefreshToken,
```
Similar comment to an above one, these seem like they might be worth their fields rather than just being in a prop bag

chenilim replied:
Agree authService should be a field (I'll update the model). @jespino, what do you think about expiry and refreshToken? They are service specific, so maybe should remain as props? Or maybe the prop store was meant for something else?

jwilander commented on server/mmoauth/mmoauth.go:
```diff
@@ -0,0 +1,248 @@
+	// TODO: Review security of this approach
+	var expiration = time.Now().Add(365 * 24 * time.Hour)
+	tokenCookie := http.Cookie{Name: "oauthtoken", Path: "/", Value: token.AccessToken, Expires: expiration, HttpOnly: false}
+	http.SetCookie(w, &tokenCookie)
```
I don't think this part needs to be specific to an oauth token. I would unify this with how non-oauth login works (which looks like via localstorage). Having two different ways of storing the authentication token on the webapp/client increase the attack surface and chance for bad security bugs IMO. And the webapp/client doesn't really need to care about whether it's oauth or not once it has it's session.

chenilim replied:
So, the `oauthtoken` cookie is just used to pass it to the client JS code, which (in app.tsx) copies it into localStorage and deletes it. This is because OAuth does a redirect to the callback, instead of going through a fetch like the login API. Do you, or does anyone else know of a better way to pass the token back to the client code?

jwilander replied:
Got it. Missed that. One way to do it is to have the redirect go to a webapp page that makes the API request to the oauth callback endpoint. That way you could still return a JSON as a response for the webapp to handle the token from like you do for regular login. Other option would be to have regular login use a cookie too. WDYT?

chenilim replied:
Thanks for the idea! I realized the callback page can just render JS to set the token and redirect, so reimplemented it that way and removed the unnecessary 'oauthtoken' cookie.

jwilander commented on webapp/src/app.tsx:
```diff
@@ -31,10 +33,27 @@ export default class App extends React.PureComponent<unknown, State> {
+        this.storeTokenFromCookie()
+    }
+
+    private storeTokenFromCookie() {
```
Per above comment I would suggest handling token storage the same way it's done for regular login rather than supporting both cookie and localstorage as options

chenilim replied:
With the sessionID set by rendered JS, this code is now gone, and it's much cleaner now.

jwilander reviewed (changes requested):
Good stuff so far 👍 Sorry about the barrage of comments 😅

chenilim commented on server/api/api.go:
```diff
@@ -101,15 +116,49 @@ func (a *API) checkCSRFToken(r *http.Request) bool {
+		WorkspaceID: workspaceID,
+	}
+
+	if workspaceID == "" || workspaceID == "0" {
```
Yeah, native auth uses workspaceID 0 for the API calls. It's more a developer-case if you switch providers. DoesUserHaveWorkspaceAccess would fail that case anyway, but this avoids a call.

chenilim commented on server/api/api.go:
```diff
@@ -101,15 +116,49 @@ func (a *API) checkCSRFToken(r *http.Request) bool {
+	ctx := r.Context()
+	session := ctx.Value("session").(*model.Session)
+	if !a.WorkspaceAuthenticator.DoesUserHaveWorkspaceAccess(session, workspaceID) {
+		return container, errors.New("Access denied to workspace")
```
Updated the function signature to return nil on error. The error was / is logged on the server, and a generic (no workspace) error is returned in any case: if the workspace is invalid or access denied.

chenilim commented on server/api/api.go:
```diff
@@ -740,10 +878,28 @@ func (a *API) handleGetWorkspace(w http.ResponseWriter, r *http.Request) {
+		}
+		workspace = &model.Workspace{
+			ID: workspaceID,
+		}
```
Currently yes. This reuses the existing workspace model, which was used to return the signup token, not used in mm auth. I envision other properties being added to the workspace in the future.

chenilim commented on server/api/auth.go:
```diff
@@ -358,9 +358,12 @@ func (a *API) attachSession(handler func(w http.ResponseWriter, r *http.Request)
+				UserID: "single-user",
+				Props: map[string]interface{}{
+					"authService": a.authService,
+				},
```
Expediency / laziness. :) Agree authService should be a field. Not sure about expiry and refreshToken though, as they are service-specific.

chenilim replied:
Updated this, moved AuthService to a field of Session.

chenilim commented on server/mmauth/mmauth.go:
```diff
@@ -0,0 +1,249 @@
+	http.SetCookie(w, &redirectCookie)
+
+	oauthState := generateSecurityToken()
+	stateCookie := http.Cookie{Name: "oauthstate", Path: "/", Value: oauthState, Expires: expiration, HttpOnly: true}
```
Good catch. I'll add a todo to add a "useSecureCookie" config setting, and also add a step to the install docs to enable it after setting up TLS.

chenilim replied:
Added a SecureCookie config setting.

jupenur commented on server/mmauth/mmauth.go:
```diff
@@ -0,0 +1,249 @@
+	redirectCookie := http.Cookie{Name: "oauthredirect", Path: "/", Value: redirectUrl, Expires: expiration, HttpOnly: true}
+	http.SetCookie(w, &redirectCookie)
+
+	oauthState := generateSecurityToken()
```
Not entirely sure what this is referring to. The state param & cookie should prevent CSRF independently of any other session handling, so looks good to me.

jwilander commented on server-config.json:
```diff
@@ -5,6 +5,7 @@
 	"dbconfig": "./focalboard.db",
 	"postgres_dbconfig": "dbname=focalboard sslmode=disable",
 	"useSSL": false,
+    "secureCookie": false,
```
This is probably ok for now but in the future we could automatically set the secure cookie attribute based on HTTP headers like we do in the MM server https://github.com/mattermost/mattermost-server/blob/9cc5089af8b7382c0170417bd2c1808662000573/app/login.go#L243

jwilander commented on webapp/src/utils.ts:
```diff
@@ -43,6 +43,14 @@ class Utils {
+
+    static deleteCookie(name: string): void {
+        document.cookie = `${name}=; expires=Thu, 01 Jan 1970 00:00:00 UTC; Max-Age=0; path=/`
+    }
```
These are unused now, right? Can probably delete them if so

chenilim replied:
Good catch. Deleted.

jwilander reviewed (approved):
Looks good 👍 there's room to improve (as always) but this is a great starting point. Left a couple non-blocking comments

chenilim commented:
Thanks for the reviews @jwilander and @jupenur! I've updated the code and refactored this PR into separate checkins. The workspaceID is [checked in here](https://github.com/mattermost/focalboard/commit/725971784b064803850d6e94777cab13788d4b7f), and I'll update the auth separately. Closing this PR.
